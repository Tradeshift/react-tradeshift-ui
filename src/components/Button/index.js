import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Button extends Component {
	/*
	 * The goal here is to add class names to the classList array
	 * while keeping the class names that the Tradeshift UI library
	 * adds after spiritualizing the element. Passing props directly to
	 * the className attribute would overwrite the existing names so
	 * we must make a reference to the DOM node and update the classList
	 * array dynamically upon passing new refs.
	 *
	 * Partially this works for the storybook application, but the
	 * idea applies for normal instantiation as well.
	 *
	 */
	componentDidMount() {
		this.props.classes.forEach(addedClass => {
			this.button.classList.add(addedClass.trim());
		});
	}

	componentWillReceiveProps({ classes }) {
		// Remove the previous type from the classList array
		// Trim for safety because it may cause `Uncaught DOMException`
		this.props.classes.forEach(addedClass => {
			this.button.classList.remove(addedClass.trim());
		});
		// Add the new one
		classes.forEach(addedClass => {
			this.button.classList.add(addedClass.trim());
		});
	}

	render() {
		const buttonProps = {
			'data-ts.busy': this.props.busy
		};

		return (
			<button
				id={this.props.id}
				classes={this.props.classes}
				data-ts="Button"
				{...buttonProps}
				// Set this ref so we can grab the classList later after
				// TS.ui spiritualizes and adds classes.
				ref={el => (this.button = el)}
			>
				{this.props.children}
			</button>
		);
	}
}

Button.propTypes = {
	busy: PropTypes.bool,
	id: PropTypes.string,
	classes: PropTypes.arrayOf(PropTypes.string),
	children: PropTypes.element.isRequired
};

Button.defaultProps = {
	busy: false,
	classes: [],
	id: '',
	children: <span>Default Button</span>
};

export default Button;
