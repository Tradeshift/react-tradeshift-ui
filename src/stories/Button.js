import { storiesOf } from '@storybook/react';
import React from 'react';
import { Button } from '../components';

const stories = storiesOf('Button', module);

stories.add(
	'Basic structure',
	() => (
		<Button>
			<span>Default Button</span>
		</Button>
	),
	{
		info: 'A simple Button does not require any props',
	},
);

stories.add(
	'Primary button',
	() => (
		<Button className="ts-primary">
			<span>Primary Button</span>
		</Button>
	),
	{
		info: 'A primary Button requires `ts-primary` className',
	},
);

stories.add(
	'Secondary button',
	() => (
		<Button className="ts-secondary">
			<span>Secondary Button</span>
		</Button>
	),
	{
		info: 'A secondary Button requires `ts-secondary` className',
	},
);

stories.add(
	'Tertiary button',
	() => (
		<Button className="ts-tertiary">
			<span>Tertiary Button</span>
		</Button>
	),
	{
		info: 'A tertiary Button requires `ts-tertiary` className',
	},
);

stories.add(
	'Micro button',
	() => (
		<Button className="ts-primary ts-micro">
			<span>Micro Button</span>
		</Button>
	),
	{
		info:
			'A micro Button requires `ts-micro` and one of `ts-primary` `ts-secondary` `ts-tertiary` className',
	},
);

stories.add(
	'Busy button',
	() => (
		<Button busy className="ts-primary">
			<span>Busy Button</span>
		</Button>
	),
	{
		info: 'For making a Button busy, set `busy={true}`',
	},
);

stories.add(
	'Icon button',
	() => (
		<Button className="ts-primary">
			<i className="ts-icon-arrowright" />
		</Button>
	),
	{
		info: 'For more information visit `tradeshift-ui` documentation page',
	},
);
