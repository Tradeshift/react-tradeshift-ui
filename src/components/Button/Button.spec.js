import React from 'react';
import { shallow, mount } from 'enzyme';

import Button from './';

describe('Button', () => {
	it('renders with default props', () => {
		const wrapper = shallow(<Button />);
		expect(wrapper).toMatchSnapshot();
	});

	describe('Button Prop Functionality', () => {
		it('loads the primary class', () => {
			const wrapper = mount(<Button />);

			wrapper.setProps({
				classes: ['ts-primary', 'ts-micro']
			});

			expect(wrapper.find('.ts-primary .ts-micro')).toBeTruthy();
		});
	});
});
