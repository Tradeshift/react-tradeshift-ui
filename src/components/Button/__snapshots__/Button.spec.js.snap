exports[`Button renders with default props 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <button
    data-classes={Array []}
    data-ts="Button"
    data-ts.busy={false}
    id="">
    <span>
        Default Button
    </span>
</button>,
  "nodes": Array [
    <button
      data-classes={Array []}
      data-ts="Button"
      data-ts.busy={false}
      id="">
      <span>
            Default Button
      </span>
</button>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Button
        busy={false}
        classes={Array []}
        id="">
        <span>
                Default Button
        </span>
</Button>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Button {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "busy": false,
          "children": <span>
            Default Button
</span>,
          "classes": Array [],
          "id": "",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <button
          data-classes={Array []}
          data-ts="Button"
          data-ts.busy={false}
          id="">
          <span>
                    Default Button
          </span>
</button>,
        "_debugID": 2,
        "_renderedOutput": <button
          data-classes={Array []}
          data-ts="Button"
          data-ts.busy={false}
          id="">
          <span>
                    Default Button
          </span>
</button>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Button
    busy={false}
    classes={Array []}
    id="">
    <span>
        Default Button
    </span>
</Button>,
}
`;
